crates/iroha_config/src/parameters/user.rs

```diff
@@ -273,6 +273,8 @@ pub struct Network {
     /// Peer-to-peer address
     #[config(env = "P2P_ADDRESS")]
     pub address: WithOrigin<SocketAddr>,
+    #[config(env = "P2P_EXTERNAL_PORT")]
+    pub external_port: WithOrigin<u16>,
```

Comment:
I think the notion of `port`, and it being a required parameter, is not the right decision.

Knowing only the port is not sufficient to connect to a peer. They might be hosted on different hosts, especially in environments like Kubernetes cluster, where different peers might have different associated domains. While `network.address` is used to _bind_ the peer to listen to a specific address in the local environment, there could be an _optional_ mention of `network.public_address` which other peers could use to connect to this peer. Moreover, in some environments it could make sense for it to be an array, so that peer will broadcast all of them to all peers and each of the peers will use the first reachable one. Being optional, peer will broadcast its local address (`network.address`) by default.

Maybe I am not right. Could you explain how `external_port` would work in a context when peers are distributed around multiple machines?

Reply:
> Could you explain how `external_port` would work in a context when peers are distributed around multiple machines?

Consider we have network with PeerA and some other peers, and add new PeerX. Initially network doesn't know PeerX address and knows only its public key. Initially PeerX knows only PeerA address.
* PeerX tries to connect to PeerA
* PeerA receives connection from some IP 1.2.3.4
* PeerA accepts the connection because it knows PeerX public key
* PeerX sends its `external_port` (5555) to PeerA
* PeerA knows PeerX IP (1.2.3.4) and its port (5555) and so forms its address (1.2.3.4:5555)
* PeerA broadcasts PeerX address (1.2.3.4:5555) to other peers

So here we assume that other peers could reach PeerX through the same IP (1.2.3.4) as the PeerA. I agree that this assumption might be incorrect. E.g. if PeerA and PeerX are both in some kubernetes cluster, PeerA might receive connection from PeerX using some local kubernetes IP. Such IP would not work for PeerB outside of the cluster.

So I like the idea of providing full address in the config (that is replace `network.external_port` with `network.public_address`). But I don't understand why it should be optional. As I understand, `network.address` usually is like `0.0.0.0:port`, so basically it contains only port to bind and can omit the host.

Reply:
> As I understand, `network.address` usually is like `0.0.0.0:port`, so basically it contains only port to bind and can omit the host.

It is not always the case. One could specify some other IP to bind to a specific network interface for security (to limit access) or other reasons. It might be crucial for advanced deployments to be able to alter the host.

But I would agree that in most cases it is `0.0.0.0`. Shall we split `network.address` to required `port` and optional `host` defaulting to `0.0.0.0`?

> So I like the idea of providing full address in the config (that is replace `network.external_port` with `network.public_address`). But I don't understand why it should be optional.

Maybe you are right and `network.public_address` is needed in the majority in cases. My rationale was that "for some setups it might not be needed and peers could broadcast their `network.address` as if it is `network.public_address`"

Reply:
@s8sato ideas here?

Reply:
> Shall we split `network.address` to required `port` and optional `host` defaulting to `0.0.0.0`?

Personally I think current parameter `network.address` is OK

> for some setups it might not be needed and peers could broadcast their `network.address` as if it is `network.public_address`

I agree that it may be convenient for some setups, but for complex setups it probably will be hard to understand for users that peers are not connecting to each other because some optional parameter is not specified. So I think it should be required parameter

Reply:
No objection from my side.
I think a joining peer X only has to notify its public address, and should't include its private address in the notification, because the private address can conflict with another peer's one in another cluster

Let's note your decision as doc comments to clarify `address` `public_address` responsibilities

Reply:
Agreed.

Also, maybe `gossip_address` would be a clearer name.

Reply:
> I think a connecting peer X has to notify its public address only i.e. should't include its private address in the notification, because the private address can conflict with another peer's one in another cluster.

Could you elaborate on this please? It is already the case that when connected, peer sends only its public address, isn't it?



> Let's note your decision as doc comments to clarify `address` `public_address` responsibilities

Added comment to `iroha_config::parameters::user::Network`

Reply:
@dima74 

> It is already the case

Yes, I just supported your current design. My intention was to oppose the idea that private address also can be broadcast